docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+    2. Call the user code `fn`
+    3. Collect the updated variables and rng and return it together with the original return values from `fn`
+3. Update the original state with the updated state returned from the transformation.
+
+A more in depth explanation of functionalization and lifting can be found [here](https://flax.readthedocs.io/en/latest/design_notes/lift.html).
+
+### Practical consequences
+
+For the most part functionalization is something that is handled automatically for you. Still there are some constraints that you must take into account. Most importantly, Flax only handles the stateful primitives (Linen variables and RNGs) and not arbitrary stateful Python code. Most importantly: You cannot close over stateful objects and `Module` objects because they are invisible to Flax's internals (and to JAX in general).
+
+
+```python
+class Foo(nn.Module):
+
+  @nn.compact
+  def __call__(self, x):
+    dense = nn.Dense(x.shape[-1])
+    return self.inner(x, lambda x: dense(x) + 1)
+  
+  def inner(self, x, fn):
+    for i in range(3):
+      x = fn(x)
+    return x
+```
+
+
+Here `inner` takes a function that closes over a Module instance. In this example, that works fine because we are not transforming the inner method with a lifted transformation. Most methods are not transformed but it is good to know how to make Module methods transformable.
+
+The main obstacle for transformability are types that JAX does not recognize. By default JAX will recognize containers (lists, dictionaries, and tuples) of JAX/NumPy arrays and python scalars (bool, int, float, etc.). The containers are called [pytrees](https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html) and the pytree API is extensible with custom types. With the flax.struct package you can also define pytree compatible dataclasses.
+
+Function closure is the most common way to accidentally hide a JAX array or linen Module from a transformation. There is however an easy workaround if you want to pass closures that are also compatible with JAX and Linen transformations:
+
+
+```python
+class Partial(flax.struct.PytreeNode):
+  fn: Callable = flax.struct.field(pytree_node=False)
+  args: Iterable[Any]
+
+  def __call__(self, *args, **kwargs):
+    return self.fn(*(tuple(self.args) + args), **kwargs)
+
+class Foo(nn.Module):
+
+  @nn.compact
+  def __call__(self, x):
+    dense = nn.Dense(x.shape[-1])
+    return self.inner(x, Partial(lambda mdl, x: mdl(x) + 1, dense)
+  
+  def inner(self, x, fn):
+    ...
+```
+
+
+Here the closure is implemented using a Flax dataclass. The function itself is annotated with `flax.struct.field(pytree_node=False)` to indicate that it does not contain Jax Arrays or linen Modules. The partially applied `args` on the other hand is treated as a pytree container. We rewrite the closure to use Partial. Now the inner method can be transformed using lifted transformations.
+
+
+## Future work
+
+
+### Setup for unbound Modules
```

Comment:
This section feels like it would be better a separate FLIP.

Not sure why this would be relevant to users reading about Linen module lifecycle. And publishing it as part of a design note will just archive the discussion, instead of discussing the pros and cons of its implementation.

Reply:
I see your point... My intention here is to give an idea about what is still being actively worked on / up for discussion. Of course, this shouldn't go in depth about the actual implementation, backward comp, etc. And if we eventually do make this change it should be part of a FLIP.